### PR TITLE
bootstrap using uv

### PR DESCRIPTION
**Pull request by dsp**
*NOTE* This is the first iteration on the PR. There is probably a bunch to clean up. Nevertheless I am happy to get early feedback and iterate.

This PR aims to remove pip-tools with uv. We do this by:

1. Create a download url generator for uv downloads.
2. Rework our generated downloads structure to encapsulate python and uv downloads.
3. Implement bootstrapping using UV using the download urls. We will always fetch the only version we have in our downloads.

At the moment we only include one version of uv in the rye binary and ensure this one is used for bootstrapping (multiple versions can co-exists in $RYE_HOME/uv)

Open questions:
1. Should we fall-back to pip-tools if uv downloads fail?
2. How to best test this? Should we make it optionally behind a feature flag `--use-uv` for a while?

### CODE REVIEW

**Comment by dsp** [comment]
Okay at least we bootstrap with pip tools and this will work in the case of `behavior.use-uv=true`. Still have to make it work for `behavior.use-uv=false`.

Docs are missing as well

**Comment by dsp** [comment]
Okay the pull request as is works with pip-tools. There is an open question regarding pip-tools handling. At the moment pip-tools bootstraps into $RYE_HOME/pip-tools since we need it to bootstrap $RYE_HOME/self. However, since we now bootstrap with uv, we could just add `pip-tools` to `SELF_REQUIREMENTS` and use that during sync when `behavior.use-uv` is false. However, while i think that's the right thing to do, it's probably best if we do this in a separate PR. Let uv bootstrapping get released and after a while change pip tools. At least have it in another PR, since we squash commits during merging, we ight want to have it separately in case we need to revert it.

**Comment by dsp** [comment]
I ran some basic local tests, both with uv bootstrapping:
1. Use uv and create a new project and sync: Works ✅
2. Use pip-tools and create a new project and sync: Works ✅

**Comment by mitsuhiko** [comment]
> However, since we now bootstrap with uv, we could just add pip-tools to SELF_REQUIREMENTS and use that during sync when behavior.use-uv is false

The issue is that pip-tools needs a fresh venv matching the target python version :( Cannot be installed into the global venv.

**Comment by dsp** [comment]
> > However, since we now bootstrap with uv, we could just add pip-tools to SELF_REQUIREMENTS and use that during sync when behavior.use-uv is false
> 
> The issue is that pip-tools needs a fresh venv matching the target python version :( Cannot be installed into the global venv.

Right. Let's keep it then how it is. There is a good chance I've missed a piece to the puzzle, but with my basic testing the PR is good for review.

File: rye/src/bootstrap.rs [hunk 1]
```diff
@@ -547,6 +466,187 @@ pub fn download_url_ignore_404(url: &str, output: CommandOutput) -> Result<Optio
...
+    uv_bin: PathBuf,
+}
+
+impl Uv {
+    // Ensure we have a uv binary for bootstrapping
+    fn ensure_exists(output: CommandOutput) -> Result<Self, Error> {
+        // Request a download for the default uv binary for this platform.
+        // For instance on aarch64 macos this will request a compatible uv version.
+        let download = UvDownload::try_from(UvRequest::default())?;
+        let uv_dir = get_app_dir().join("uv").join(download.version());
```
Comment (mitsuhiko) [inline rye/src/bootstrap.rs:483]: Is there much of a benefit of having multiple uv versions and not just to retain only the latest?

**Review by mitsuhiko — CHANGES_REQUESTED** [review]
The `uv==0.1.9` line in the self requirements at this point probably should be removed. Otherwise you end up with two uvs for no good reason.

**Comment by mitsuhiko** [comment]
Other than the few nits I think this looks good. Also makes for a quicker bootstrap process. I think a good followup would be to make the virtualenv quiet as it now spits out some quite confusing output during installation:
 
```
Downloaded cpython@3.12.1
Using Python 3.12.1 interpreter at C:\Users\armin\.rye\py\cpython@3.12.1\install\python.exe
Creating virtualenv at: C:\Users\armin\.rye\self
Activate with: C:\Users\armin\.rye\self\Scripts\activate
Resolved 1 package in 2ms
Installed 1 package in 95ms
 + pip==23.3.2
Resolved 35 packages in 24ms
```

Since that env really only exists internally I don't think it makes sense to reveal where it is and how it's being activated.

File: rye/src/bootstrap.rs (hunk 1 above)
Comment (dsp) [reply rye/src/bootstrap.rs:483]: The main reason here is that i can check for the version with one simple `stat` call and not have to `execv` uv. This allows me in later changes to make use of `ensure_exists` to obtain a `Uv` on any code path. It also opens the door for pinning uv if we ever need to (I don't see why, but it feels more future proof). I am happy to just nuke the uv directory before I download a new version which will keep it clean for now. If you insist i can remove the version.

**Comment by dsp** [comment]
> The `uv==0.1.9` line in the self requirements at this point probably should be removed. Otherwise you end up with two uvs for no good reason.

Okay that makes sense. This will add a few more changes as we rely on uv being in `.rye/self/bin`. I'll follow up with adjustment to path we use. A later changset will make use of `Uv` everywhere where we currently check for Config::current().use_uv() and then construct path to the binary. This should be more sane and allows us to have a canoncial, encapsulated way to construct the binary path. It will also allow us to ensure on a type level that in certain contexts we always operate on a venv.

**Comment by dsp** [comment]
Okay changes made: I keep `uv/[version]/uv` for now but will clean old directories after we downloaded and extracted the new version. This feels to me a bit more future proof and allows nicely to inspect which uv version we are on.

**Comment by mitsuhiko** [comment]
Looks good!

**Comment by bluss** [comment]
@ImportTaste what about all other non-relocatable things in a python virtualenv, is there a solution for them?

Let's say I inspect for example ~/.rye/self/bin/pip3 and find that its shebang has an absolute path as well to the python in the venv, this is typically how virtual environments work unfortunately(!). But since it's a long standing issue, I could see there could be tools to "relocate" them.

**Pull request merged**